conversion/resource_generation.go

```diff
@@ -0,0 +1,401 @@
+	if err != nil {
+		err = fmt.Errorf("unable to create connection profiles: %v", err)
+		return err
+	}
+	conv.ResourceValidation = true
+	return nil
+}
+
+// 1. If destination connection profile needs to be created, creates a gcs bucket
+// 2. Creates the connection profile needed for migration
+func (r ResourceGenerationImpl) PrepareMinimalDowntimeResources(createResourceData *ConnectionProfileReq, mutex *sync.Mutex) common.TaskResult[*ConnectionProfileReq] {
+	req := &datastreampb.CreateConnectionProfileRequest{
+		Parent:              fmt.Sprintf("projects/%s/locations/%s", createResourceData.ConnectionProfile.ProjectId, createResourceData.ConnectionProfile.Region),
+		ConnectionProfileId: createResourceData.ConnectionProfile.Id,
+		ConnectionProfile: &datastreampb.ConnectionProfile{
+			DisplayName:  createResourceData.ConnectionProfile.Id,
+			Connectivity: &datastreampb.ConnectionProfile_StaticServiceIpConnectivity{},
+		},
+		ValidateOnly: createResourceData.ConnectionProfile.ValidateOnly,
+	}
+
+	// If destination source profile is to be created, create a gcs bucket first
+	var bucketName string
+	if !createResourceData.ConnectionProfile.IsSource {
+		bucketName = strings.ToLower("GCS-" + createResourceData.ConnectionProfile.Id)
+		err := r.StorageAcc.CreateGCSBucket(createResourceData.Ctx, r.StorageClient, storageaccessor.StorageBucketMetadata{
+			BucketName: bucketName,
+			ProjectID:  createResourceData.ConnectionProfile.ProjectId,
+			Location:   createResourceData.ConnectionProfile.Region,
+		})
+		if err != nil {
+			createResourceData.Error = err
+			return common.TaskResult[*ConnectionProfileReq]{Result: createResourceData, Err: err}
+		}
+	}
+	createResourceData.ConnectionProfile.BucketName = bucketName
+
+	// Set Profile for resource creation
+	setConnectionProfileFromRequest(createResourceData, req)
+
+	// Create or Validate Resource
+	_, err := r.DsAcc.CreateConnectionProfile(createResourceData.Ctx, r.DsClient, req)
+	if err != nil {
+		createResourceData.Error = err
+		return common.TaskResult[*ConnectionProfileReq]{Result: createResourceData, Err: err}
+	}
+
+	if !createResourceData.ConnectionProfile.ValidateOnly {
+		fmt.Printf("Connection Profile for Datashard %v has been created: %v\n", createResourceData.ConnectionProfile.DatashardId, createResourceData.ConnectionProfile.Id)
+		// In case of failure, add resources to be cleaned up
+		resourcesForCleanup = append(resourcesForCleanup, createResourceData)
+	} else {
+		fmt.Printf("Connection Profile for Datashard %v has been validated: %v\n", createResourceData.ConnectionProfile.DatashardId, createResourceData.ConnectionProfile.Id)
+	}
+
+	return common.TaskResult[*ConnectionProfileReq]{Result: createResourceData, Err: nil}
+}
+
+// If any of the resource creation fails, deletes all resources that were created
+func (r ResourceGenerationImpl) ConnectionProfileCleanUp(ctx context.Context, profiles []*ConnectionProfileReq) error {
```

Comment:
You should name this something like `RollbackResourceCreation` or similar to demarcate what this operation is for (and not what it does)